### ISSUE TITLE
Passing functions

### ISSUE THREAD

**Issue by hiptmair** [craffael/lehrfempp#76]
Often functions take functors as arguments that are supposed to pass objects with an evaluation operator _that can also be lambda functions_. What is the right way to pass such objects:
1.
~~~
template<FUNCTOR> foo(FUNCTOR func); 
~~~~
2.
~~~
template<FUNCTOR> foo(FUNCTOR &func); 
~~~~
3.
~~~
template<FUNCTOR> foo(FUNCTOR &&func); 
~~~~

Passing as `const` is not desirable, since the object, may, e.g., count the number of invocations.

**Comment by craffael** [comment]
I would usually use 3) but let me comment a bit on this:

1. has the effect that `func` is copied/moved whenever `foo` is called. So if `func` changes it's internal state everytime it is called and you want to retrieve the final state after calling `foo`, copying `func` is the wrong approach. Contrary to 2), 1) is able to deal with r-values, e.g.`foo([](){return "bla";})` is valid.

2. is a rather bad approach because it cannot bind to r-valudes: `foo([](){return "bla";})` is not allowed.

3. Supports l-values and r-values and `func` is always passed with reference semantics. This means in particular that `foo` can modify `func`.

**Comment by craffael** [comment]
yes of course, rule 3) from above only works the way I've explained it, if the function/method is a function template:
```
// here f is a universal reference, accepts l-values and r-values
template<class FUNCTOR> foo(FUNCTOR&& f);

template<class FUNCTOR>
class Foo {
  Foo(FUNCTOR&& f); // if FUNCTOR is the type of a lambde, e.g. decltype([](){return 0;}), then this constructor accepts only r-values
};
```
If you want to have a behavior as in 3), you could write it up like this:
```
template<class FUNCTOR>
class Foo {
  // accept r-values
  Foo(FUNCTOR&& f);
  // accept l-values:
  Foo(const FUNCTOR& f);
};
```
